File: codeforces/427C-checkposts/main.cpp
#include <iostream>
#include <vector>
#include <functional>
#include <algorithm>
typedef long long LL;
std::vector<std::vector<int>> tarjanSCC(std::vector<std::vector<int>> &adj) {
	int n = adj.size(), clock = 0;
	std::vector<int> low(n), dis(n); // 'dis'covery time and 'low'est node reachable
	std::vector<int> has(n), stk; // 'has' checks if 'stk' has a node
	std::vector<std::vector<int>> sccs; // holds a list of sccs with their nodes
	std::function<void (int)> dfs = [&](int u) {
		dis[u] = low[u] = ++clock;
		stk.push_back(u);
		has[u] = true;
		for (int v : adj[u]) {
			if (!dis[v])
				dfs(v);
			if (has[v]) // On the stack means found lower
				low[u] = std::min(low[u], low[v]);
		}
		if (low[u] == dis[u]) { // No path to low nodes, so it's the cutoff for an SCC
			sccs.push_back({});
			while (true) {
				int v = stk.back();
				stk.pop_back();
				has[v] = false;
				sccs.back().push_back(v);
				if (u == v)
					break;
			}
		}
	};
	for (int u = 0; u < adj.size(); ++u) // assumes zero indexed
		if (!dis[u])
			dfs(u);
	return sccs;
}
int main() {
	int n, m;
	std::cin >> n;
	std::vector<int> cost(n);
	for (int i = 0; i < n; ++i)
		std::cin >> cost[i];
	std::cin >> m;
	std::vector<std::vector<int>> adj(n);
	for (int i = 0, u, v; i < m; ++i)
		std::cin >> u >> v, adj[u-1].push_back(v-1);
	auto sccs = tarjanSCC(adj);
	LL ans = 0, ways = 1, mod = 1e9+7;
	for (auto &scc : sccs) {
		int min_cost = cost[scc[0]], total = 0;
		for (int u : scc)
			min_cost = std::min(min_cost, cost[u]);
		for (int u : scc)
			if (cost[u] == min_cost)
				++total;
		ans += min_cost;
		ways = (ways*total) % mod;
	}
	std::cout << ans << ' ' << ways << '\n';
}

File: codeforces/427C-checkposts/main.py
def tarjanSCC(adj: [[int]]):
	n, clock = len(adj), iter(range(1, len(adj)+1))
	dis, low = [0]*n, [0]*n # 'dis'covery time and 'low'est node reachable
	has, stk = [0]*n, [] # 'has' checks if a node is on the DFS 'stk'
	sccs: [[int]] = [] # holds a list of sccs with their nodes
	def dfs(u: int):
		dis[u] = low[u] = next(clock)
		stk.append(u)
		has[u] = True
		for v in adj[u]:
			if not dis[v]:
				dfs(v)
			if has[v]: # on the stack means found lower
				low[u] = min(low[u], low[v])
		if low[u] == dis[u]: # did not find lower
			sccs.append([])
			while True:
				v = stk.pop()
				has[v] = False;
				sccs[-1].append(v)
				if v == u:
					break
	for u in range(n):
		if not dis[u]:
			dfs(u)
	return sccs

n = int(input())
cost = list(map(int, input().split()))
adj = [[] for _ in range(n)]
m = int(input())
for _ in range(m):
	u, v = map(int, input().split())
	adj[u-1].append(v-1)
sccs = tarjanSCC(adj)
ans, ways, mod = 0, 1, 10**9+7
for scc in sccs:
	min_elem, total = cost[scc[0]], 0
	for u in scc:
		min_elem = min(min_elem, cost[u])
	for u in scc:
		if cost[u] == min_elem:
			total += 1
	ans += min_elem
	ways = (ways*total) % mod
print(ans, ways)
